### ISSUE TITLE
Object of type "None" cannot be called

### ISSUE THREAD

**Issue by FindDefinition** [microsoft/pyright#3737]
## Environment data

-   Language Server version: 2022.7.40
-   OS and version: Ubuntu 20.04
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8

## Code Snippet

```python
from typing import Any, Callable, List, Optional

def main():
    datas: List[Any] = []
    for data in datas:
        cb: Optional[Callable[[str], None]] = None
        if cb is not None:
            lambda: cb(data)
```

## Repro Steps

1. Paste code above to VSCode

## Expected behavior
no error reported
## Actual behavior
pylance report a error

## Logs

![problem](https://user-images.githubusercontent.com/73811981/180600120-264dd0df-9e3b-47e5-b09a-9708450ebb76.jpg)

**Comment by erictraut** [comment]
This is expected behavior. You're using `cb` in a different scope (in this case, a lambda). It is unsafe in general to apply a narrowed type from the outer scope because the value of the captured variable can be modified before the lambda is invoked.

Pyright does have special code that allows the use of a narrowed type in circumstances where the captured variable used in the inner scope is unmodified at all points beyond the definition of the inner scope, such as in this case:
```python
def foo(cb: Optional[Callable[[str], None]] = None):
    if cb is not None:
        lambda: cb("")
```

But in your example above, `cb` is modified after the lambda definition because it is modified within a loop.

The suggested way to handle this is through the use of a separate variable that can be captured by the inner scope.

```python
        if cb is not None:
            temp = cb
            lambda: temp(data)
```
